### PR TITLE
Avoid duplicate OBJECT_IDs after MVU

### PR DESCRIPTION
**Pull request by 2jungkook**
Signed-off-by: Jungkook Lee <jungkook@amazon.com>

### Description

PG allows duplicated OIDs in different catalog tables. However, in Babelfish, we use PG's OID as OBJECT_ID in all system catalog views and it should be unique within a logical database (https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/relational-databases/system-catalog-views/sys-all-objects-transact-sql.md). Duplicate OIDs can break SQL server tools, e.g., SSMS, as well as customer applications.

In PG/Babelfish, duplicate OIDs can occur in the following two cases:
1. After MVU (Major Version Upgrade)
2. After hitting OID wraparound

This commit is to prevent duplicate OIDs during MVU and consists of two parts:

> First of all, from the old cluster, it finds the max OID in the 5 catalog tables: `pg_extension`, `pg_authid`, `pg_enum`, `pg_class` and `pg_type`. And then pass the max OID to the new cluster as a form of GUC, `babelfishpg_tsql.dump_restore_min_oid`. In the new cluster, set the nextOid greater than the max OID from the old cluster using `GetNewObjectId_hook`.

> Secondly, in `pg_upgrade.c`, it skips the step of transferring the next OID from the old cluster to the new cluster.

> Also note that the type of babelfishpg_tsql.dump_restore_min_oid is defined by `DefineCustomStringVariable()` to support UINT_MAX.

Also, this commit contains minor improvements on the pg-upgrade Github action and `dev-tools.sh`:
- pg-upgrade Github action: add `retain` option to see pg_upgrade log even if pg_upgrade succeeds and jdbc tests fails.
- dev-tools.sh: developers can choose the migration mode, e.g., `./dev-tools.sh test input multi-db`.

Engine PR: https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/54
 
### Issues Resolved

BABEL-3613

### Test Scenarios Covered ###
* **Use case based -** Yes
* **Boundary conditions -** Yes (babelfishpg_tsql.dump_restore_min_oid)
* **Arbitrary inputs -** N/A
* **Negative test cases -** N/A
* **Minor version upgrade tests -** Yes
* **Major version upgrade tests -** Yes
* **Performance tests -**N/A
* **Tooling impact -** N/A
* **Client tests -** N/A

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 1]
```diff
@@ -281,6 +288,19 @@ UninstallExtendedHooks(void)
...
+pltsql_GetNewObjectId(VariableCache variableCache)
+{
+	if (!babelfish_dump_restore)
+		return;
+
+	if (ShmemVariableCache->nextOid >= babelfish_dump_restore_min_oid)
+		return;
+
+	ShmemVariableCache->nextOid = babelfish_dump_restore_min_oid;
+	ShmemVariableCache->oidCount = 0;
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/hooks.c:305]: what is the implication of setting oid count to 0

Comment (2jungkook) [reply contrib/babelfishpg_tsql/src/hooks.c:305]: That is for `XLogPutNextOid`.

File: contrib/babelfishpg_tsql/src/guc.c [hunk 2]
```diff
@@ -1049,6 +1050,15 @@ define_custom_variables(void)
 				 GUC_NOT_IN_SAMPLE | GUC_DISALLOW_IN_FILE | GUC_DISALLOW_IN_AUTO_FILE,
 				 NULL, NULL, NULL);
 
+	DefineCustomIntVariable("babelfishpg_tsql.dump_restore_min_oid",
+				 gettext_noop("All new OIDs should be greater than this number during dump and restore"),
+				 NULL,
+				 &babelfish_dump_restore_min_oid,
+				 FirstNormalObjectId, 0, INT_MAX,
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/guc.c]: is INT_MAX for unsigned or signed? we should make sure it matches max oid value

File: .github/composite-actions/run-pg-upgrade/action.yml [hunk 3]
```diff
@@ -27,7 +27,7 @@ runs:
         export PATH="$HOME/${{ inputs.pg_new_dir }}/bin:$HOME/${{ inputs.pg_old_dir }}/bin:$PATH"
         export LD_LIBRARY_PATH="$HOME/${{ inputs.pg_new_dir }}/lib:$HOME/${{ inputs.pg_old_dir }}/lib:$LD_LIBRARY_PATH"
         ~/${{ inputs.pg_new_dir }}/bin/pg_upgrade -U runner -b ~/${{ inputs.pg_old_dir }}/bin -B ~/${{ inputs.pg_new_dir }}/bin \
-        -d ~/${{ inputs.pg_old_dir }}/data -D ~/${{ inputs.pg_new_dir }}/data -p 5432 -P 5433 -j 4 --link --verbose
+        -d ~/${{ inputs.pg_old_dir }}/data -D ~/${{ inputs.pg_new_dir }}/data -p 5432 -P 5433 -j 4 --link --verbose --retain
```
Comment (surendravishnoi) [inline .github/composite-actions/run-pg-upgrade/action.yml:30]: why this change?

Comment (rishabhtanwar29) [reply .github/composite-actions/run-pg-upgrade/action.yml:30]: I think we can remove `--retain` option as it retains log/sql files after successful completion but we are not even uploaded these files into artifacts if pg_upgrade completes successfully.

**Comment by rishabhtanwar29** [comment]
Let's include new test template in the PR description.

File: contrib/babelfishpg_tsql/src/guc.c (hunk 2 above)
Comment (2jungkook) [reply contrib/babelfishpg_tsql/src/guc.c]: Good catch! Will change it to UINT_MAX.

File: .github/composite-actions/run-pg-upgrade/action.yml (hunk 3 above)
Comment (2jungkook) [reply .github/composite-actions/run-pg-upgrade/action.yml:30]: Without this, we cannot see pg_upgrade log when JDBC test fails after pg_upgrade succeeds.

**Review by rishabhtanwar29 — APPROVED** [review]

**Comment by rishabhtanwar29** [comment]
Looks good to me!

**Review by kuntalghosh — APPROVED** [review]
